mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -34,72 +42,94 @@ impl BridgeShutdownHandle {
+                        device_id,
+                        settings.name()
+                    )),
+                ))
+                .with_rules(settings.forwards());
+            })
+            .with_remote(|pump| {
+                pump.with_config(MqttClientConfig::new(
+                    settings.address(),
+                    settings.keep_alive(),
+                    settings.clean_session(),
+                    settings.credentials().clone(),
+                ))
+                .with_rules(settings.subscriptions());
+            })
+            .with_store(|| PublicationStore::new_memory(BATCH_SIZE))
+            .build()?;
+
+        // TODO move susbscriptions into run method
+        local_pump
             .subscribe()
             .instrument(info_span!("pump", name = "local"))
             .await?;
 
-        pumps
-            .remote_pump
+        remote_pump
             .subscribe()
             .instrument(info_span!("pump", name = "remote"))
             .await?;
 
-        debug!("created {} bridge...", connection_settings.name());
+        debug!("created bridge...");
+
         Ok(Bridge {
-            pumps,
-            connection_settings,
+            local_pump,
+            remote_pump,
+            connection_settings: settings,
         })
     }
 
-    pub async fn run(mut self) -> Result<(), BridgeError> {
-        info!("starting {} bridge...", self.connection_settings.name());
-
-        let (local_shutdown, local_shutdown_listener) = oneshot::channel::<()>();
-        let (remote_shutdown, remote_shutdown_listener) = oneshot::channel::<()>();
-        let shutdown_handle = BridgeShutdownHandle {
-            local_shutdown,
-            remote_shutdown,
-        };
+    pub async fn run(self) -> Result<(), BridgeError> {
+        info!("Starting {} bridge...", self.connection_settings.name());
 
         let local_pump = self
-            .pumps
             .local_pump
-            .run(local_shutdown_listener)
+            .run()
             .instrument(info_span!("pump", name = "local"));
 
         let remote_pump = self
-            .pumps
             .remote_pump
-            .run(remote_shutdown_listener)
+            .run()
             .instrument(info_span!("pump", name = "remote"));
-        pin_mut!(local_pump, remote_pump);
 
         debug!(
             "starting pumps for {} bridge...",
             self.connection_settings.name()
         );
-        match select(local_pump, remote_pump).await {
-            Either::Left(_) => {
-                shutdown_handle.shutdown().await?;
+
+        select! {
+            _ = local_pump => {
+                // TODO shutdown remote pump
+                // shutdown_handle.shutdown().await?;
             }
-            Either::Right(_) => {
-                shutdown_handle.shutdown().await?;
+
+            _ = remote_pump => {
+                // TODO shutdown local pump
+                // shutdown_handle.shutdown().await?;
```

Comment:
So we have the ingress / egress shutdown handles internally within the pump, but don't yet have a exposed unifying shutdown handle to call the internal handles?

Reply:
we have `PumpMessage::Shutdown` event which I anticipated to be used to stop pumps. work to implement shutdown procedure and make tests is to be done.